### PR TITLE
lib/errors: go-multierror multi-error printing, update cockroach errors

### PR DESCRIPTION
**Pull request by bobheadxi**
After [many tangents](https://sourcegraph.slack.com/archives/C07KZF47K/p1645145091483399) this PR is a wholesale migration away from `go-multierror` into a custom multierror implementation that is fully compatible with `cockroachdb/errors`, prints all errors, can be introspected with `Is`, `As`, and friends, and more. The new multi error type is only available as an interface.

- more idiomatic way to get multi errors: `errors.As` into the `MultiError` interface, then use the read-only `Error()`
- removes the need for `ErrorOrNil`! **All multi-errors can be treated transparently as an `error` with no special handling** - you can always rely on `err != nil` to be trustworthy
- turns `CombineErrors` into multi errors as well, because the behaviour of it appears quite deceptive.

It also gets rid of the dozen or so ways things are joined into multi errors:

- no more appending directly to `err.Errors`
- no more magic `_ = Append(multi, err)` - we don't modify the provided errors anymore

This PR also removes a few error list formatters into custom error types, which is probably what they should have been anyway.

There is more discussion and background in https://github.com/sourcegraph/sourcegraph/pull/31431 . That PR is also more or less an issue description at this point, so closes https://github.com/sourcegraph/sourcegraph/pull/31431

All tests pass with only minimal changes due to removal of newlines, because we don't have `go-multierror`'s random extra `\n\n` anymore

tldr: more `error`, less `*MultiError` and `ErrorOrNil()`, better printing and consistency.

## Test plan

All tests still pass! `main-dry-run`: https://buildkite.com/sourcegraph/sourcegraph/builds/132460

Extensive-ish unit tests on the custom error type

### CODE REVIEW

**Review by tsenart — APPROVED** [review]
So good!

File: internal/observation/observation.go [hunk 1]
```diff
@@ -202,21 +202,21 @@ func (f FinishFunc) OnCancel(ctx context.Context, count float64, args Args) {
...
 	if err != nil && *err != nil {
-		e.multi.Errors = append(e.multi.Errors, *err)
+		e.errs = errors.Append(e.errs, *err)
 		e.extraFields = append(e.extraFields, fields...)
 	}
 }
 
 func (e *ErrCollector) Error() string {
-	return e.multi.Error()
+	return e.errs.Error()
```
Comment (efritz) [inline internal/observation/observation.go:219]: Can this panic?

File: lib/errors/group.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
+package errors
+
+import "sync"
+
+// Group is a collection of goroutines which return errors that need to be
+// coalesced.
+//
+// Implementation and tests are based on https://sourcegraph.com/github.com/hashicorp/go-multierror/-/blob/group.go
+type Group struct {
```
Comment (efritz) [inline lib/errors/group.go:9]: We also use [sync/errorgroup](https://pkg.go.dev/golang.org/x/sync/errgroup). Should we prefer one of them and blacklist the other?

File: lib/errors/errors.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+// CombineErrors returns a MultiError from err1 and err2. If both are nil, nil is returned.
+func CombineErrors(err1, err2 error) MultiError {
+	if err1 == nil && err2 == nil {
+		return nil
+	}
+	var multi MultiError
+	if err1 == nil {
+		if As(err2, &multi) {
```
Comment (camdencheek) [inline lib/errors/errors.go:54]: Same as my comment above, this `As` will throw away wrapping information.

Comment (camdencheek) [reply lib/errors/errors.go:54]: Actually, why do we wrap in a `MultiError` here at all? Couldn't we just return `err2` here, and have this function return `error` rather than `MultiError`?

File: lib/errors/cockroach.go [hunk 4]
```diff
@@ -19,7 +23,37 @@ var (
...
+		p.Printf("%v", err)
+	}
+
+	// Print additional details
+	if p.Detail() {
+		p.Print("-- details follow")
+		for i, err := range e.errs {
+			p.Printf("\n(%d) %+v", i+1, err)
+		}
+	}
```
Comment (efritz) [inline lib/errors/cockroach.go:56]: Curious what this output looks like.

File: lib/batches/batch_spec_test.go [hunk 5]
```diff
@@ -202,11 +193,11 @@ changesetTemplate:
 		wantErr := `3 errors occurred:
 	* on.0: Must validate one and only one schema (oneOf)
 	* on.0: Must validate at least one schema (anyOf)
-	* on.0: Must validate one and only one schema (oneOf)
-
-`
+	* on.0: Must validate one and only one schema (oneOf)`
 		haveErr := err.Error()
 		if haveErr != wantErr {
+			println(wantErr)
```
Comment (efritz) [inline lib/batches/batch_spec_test.go:199]: Keeping in?

File: lib/errors/errors.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+			multi = CombineErrors(multi, e)
+		}
+	}
+	return multi
+}
+
+func (e *multiError) Error() string { return fmt.Sprintf("%v", e) }
+func (e *multiError) Errors() []error {
+	if e == nil || e.errs == nil {
+		return []error{}
```
Comment (camdencheek) [inline lib/errors/errors.go:82]: No need for the additional `makeslice` call

```suggestion
		return nil
```

File: lib/errors/errors.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+	}
+	return e.errs
+}
+
+func (e *multiError) Cause() error  { return e.errs[len(e.errs)-1] }
+func (e *multiError) Unwrap() error { return e.errs[len(e.errs)-1] }
```
Comment (camdencheek) [inline lib/errors/errors.go:88]: This doesn't look right. I would expect `Unwrap()` to eventually hit every error in the multierror. This only returns the last error in the slice. Otherwise, `errors.Is(err1, Append(err1, err2))` will return `false`, right? Something like this:

```suggestion
func (e *multiError) Unwrap() error { return &multiError{errs: e.errs[1:] }
```

**Review by camdencheek — APPROVED** [review]
I'm very excited for this. Thank you for taking the time to do this! I love the new multierror interface. So much less awkward.

File: lib/errors/group.go (hunk 2 above)
Comment (camdencheek) [reply lib/errors/group.go:9]: They have different semantics. `sync/errorgroup` cancels immediately and returns the first error. `errors.Group` runs all subtasks to completion and returns the combined errors from all of them.

Comment (efritz) [reply lib/errors/group.go:9]: Good call though

File: internal/search/alert/observer_test.go [hunk 8]
```diff
@@ -19,22 +19,22 @@ import (
...
 	cases := []struct {
 		name                 string
-		multiErr             *errors.MultiError
+		multiErr             error
 		wantAlertDescription string
 	}{
 		{
 			name:                 "diff_search_warns_on_repos_greater_than_search_limit",
-			multiErr:             errors.Append(&errors.MultiError{}, &commit.RepoLimitError{ResultType: "diff", Max: 50}),
+			multiErr:             errors.Append(nil, &commit.RepoLimitError{ResultType: "diff", Max: 50}),
```
[the remark this answers is not in the document]
Comment (camdencheek) [reply internal/search/alert/observer_test.go:27]: I don't think there is a way to construct a multierror from outside the package without an `Append`, right?

File: lib/errors/group.go (hunk 2 above)
Comment (camdencheek) [reply lib/errors/group.go:9]: I think a unified `groups` package could be very useful. We also have a huge mishmash of techniques for rate-limiting/semaphores.

Comment (efritz) [reply lib/errors/group.go:9]: We have `internal/goroutine` already which groups could be a part of. We have parallel utilities already, we should attach an error return to one of them.

Comment (bobheadxi) [reply lib/errors/group.go:9]: This might be better for a follow-up - this PR already makes quite a lot of sweeping changes. This `Group` is mostly a copy-pasta for users of the `go-multierror` group

File: lib/errors/cockroach.go (hunk 4 above)
Comment (bobheadxi) [reply lib/errors/cockroach.go:56]:
```sh
$ go test -timeout 30s -run ^TestMultipleErrorPrinting$ github.com/sourcegraph/sourcegraph/lib/errors -v
# ...
=== RUN   TestMultipleErrorPrinting/Fancy_stack_%+v
    errors_test.go:35: deep!: hello world: 2 errors occurred:
        (1) attached stack trace
          -- stack trace:
          | github.com/sourcegraph/sourcegraph/lib/errors.TestMultipleErrorPrinting
          |     /Users/robertlin/Projects/sourcegraph/sourcegraph/lib/errors/errors_test.go:32
          | testing.tRunner
          |     /usr/local/Cellar/go/1.17.2/libexec/src/testing/testing.go:1259
        Wraps: (2) deep!
        Wraps: (3) attached stack trace
          -- stack trace:
          | github.com/sourcegraph/sourcegraph/lib/errors.TestMultipleErrorPrinting
          |     /Users/robertlin/Projects/sourcegraph/sourcegraph/lib/errors/errors_test.go:32
          | [...repeated from below...]
        Wraps: (4) hello world
        Wraps: (5) 2 errors occurred:
          |     * foo
          |     * bar
          | -- details follow
          | (1) foo
          | (1) attached stack trace
          |   -- stack trace:
          |   | github.com/sourcegraph/sourcegraph/lib/errors.TestMultipleErrorPrinting
          |   |         /Users/robertlin/Projects/sourcegraph/sourcegraph/lib/errors/errors_test.go:16
          |   | testing.tRunner
          |   |         /usr/local/Cellar/go/1.17.2/libexec/src/testing/testing.go:1259
          |   | runtime.goexit
          |   |         /usr/local/Cellar/go/1.17.2/libexec/src/runtime/asm_amd64.s:1581
          | Wraps: (2) foo
          | Error types: (1) *withstack.withStack (2) *errutil.leafError
          | (2) bar
          | (1) attached stack trace
          |   -- stack trace:
          |   | github.com/sourcegraph/sourcegraph/lib/errors.TestMultipleErrorPrinting
          |   |         /Users/robertlin/Projects/sourcegraph/sourcegraph/lib/errors/errors_test.go:17
          |   | testing.tRunner
          |   |         /usr/local/Cellar/go/1.17.2/libexec/src/testing/testing.go:1259
          |   | runtime.goexit
          |   |         /usr/local/Cellar/go/1.17.2/libexec/src/runtime/asm_amd64.s:1581
          | Wraps: (2) bar
          | Error types: (1) *withstack.withStack (2) *errutil.leafError
        Wraps: (6) attached stack trace
          -- stack trace:
          | github.com/sourcegraph/sourcegraph/lib/errors.TestMultipleErrorPrinting
          |     /Users/robertlin/Projects/sourcegraph/sourcegraph/lib/errors/errors_test.go:17
          | testing.tRunner
          |     /usr/local/Cellar/go/1.17.2/libexec/src/testing/testing.go:1259
          | runtime.goexit
          |     /usr/local/Cellar/go/1.17.2/libexec/src/runtime/asm_amd64.s:1581
        Wraps: (7) bar
        Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *withstack.withStack (4) *errutil.withPrefix (5) *errors.multiError (6) *withstack.withStack (7) *errutil.leafError
```

File: lib/errors/errors.go (hunk 3 above)
Comment (bobheadxi) [reply lib/errors/errors.go:54]: We must provide a multi error because a lot of callers of `Append` depend on a `MultiError` type being returned. I did end up changing most of those, but this makes it easier to work with for the code that knows this is going to be a multi error. If there is a single error the multierror "acts" like a single error - it prints nothing but a single error, etc.

File: lib/errors/errors.go (hunk 7 above)
Comment (camdencheek) [reply lib/errors/errors.go:85]: Actually, this doesn't really do it either. However, I think `Unwrap()` is only really used by stdlib when trying to do `errors.Is` and `errors.As`, so if we provide a custom `As` (in addition to the custom `Is` below), I think this would fix my concern.

File: lib/errors/cockroach.go (hunk 4 above)
Comment (efritz) [reply lib/errors/cockroach.go:56]: Are errors ever not details with maybe less information?

File: lib/errors/errors.go (hunk 7 above)
Comment (bobheadxi) [reply lib/errors/errors.go:83]: I think you are right, thought curiously without this change the following still works:

```
		{
			name:        "Append",
			err:         Append(errFoo, errBar),
			wantStrings: []string{"foo", "bar"},
			wantIs:      []error{errFoo, errBar},
		},
```

Comment (camdencheek) [reply lib/errors/errors.go:83]: I think this is because we provide a custom `Is` below, so calling `errors.Is()` doesn't use `Unwrap()` to get the underlying error. However, I expect `errors.As` doesn't work since we don't (yet) provide a custom implementation.

Comment (camdencheek) [reply lib/errors/errors.go:83]: A reproducing test for ya:
```go
type testError1 struct{}

func (t *testError1) Error() string { return "foo" }

type testError2 struct{}

func (t *testError2) Error() string { return "bar" }

func TestAs(t *testing.T) {
	err1 := &testError1{}
	err2 := &testError2{}
	multi := Append(err1, err2)
	var e1 *testError1
	if !As(multi, &e1) {
		t.Fatalf("expected as to work")
	}
}
```

Comment (camdencheek) [reply lib/errors/errors.go:83]: ^^ note that the above test _does_ work for `testError2` because we unwrap to the last error in the slice

Comment (bobheadxi) [reply lib/errors/errors.go:83]: > A reproducing test for ya:

Amending d28374e15826c4d2465aa98effaf0dc73b5f1c95 to look more like that I can reproduce the issue 🤔

Comment (camdencheek) [reply lib/errors/errors.go:83]: Do the tests you added actually exercise `As`? I don't see any calls to `As` anywhere

Comment (bobheadxi) [reply lib/errors/errors.go:83]: > Do the tests you added actually exercise As? I don't see any calls to As anywhere

Yeah my bad I did some funky undo - I think it was actually failing, https://github.com/sourcegraph/sourcegraph/commit/d44f7dea994a68db28965ad60841b9a93c4b7091 is the updated test

Comment (camdencheek) [reply lib/errors/errors.go:83]: Also, I agree that my original suggestion is wrong. It assumed that we didn't implement a custom `Is` (and even then it wouldn't have worked), but we do. I think all that's needed for my test case to work as expected is a custom `As` method on `multiError` similar to our custom `Is` method.

Comment (bobheadxi) [reply lib/errors/errors.go:83]: We're close-ish, but the `Append(Append(Append(Append))` test case fails, with the second append failing to assert errBaz:

```
        /Users/robertlin/Projects/sourcegraph/sourcegraph/lib/errors/errors_test.go:87: 4 errors occurred:
            (1) 4 errors occurred:
              | 	* foo
              | 	* bar
              | 	* baz
              | 	* zoo
              | -- details follow
              | (1) foo
              | (1) attached stack trace
              |   -- stack trace:
              |   | github.com/sourcegraph/sourcegraph/lib/errors.TestMultiError
              |   | 	/Users/robertlin/Projects/sourcegraph/sourcegraph/lib/errors/errors_test.go:22
              |   | testing.tRunner
              |   | 	/usr/local/Cellar/go/1.17.2/libexec/src/testing/testing.go:1259
              |   | runtime.goexit
              |   | 	/usr/local/Cellar/go/1.17.2/libexec/src/runtime/asm_amd64.s:1581
              | Wraps: (2) foo
              | Error types: (1) *withstack.withStack (2) *errutil.leafError
              | (2) bar
              | (1) attached stack trace
              |   -- stack trace:
              |   | github.com/sourcegraph/sourcegraph/lib/errors.TestMultiError
              |   | 	/Users/robertlin/Projects/sourcegraph/sourcegraph/lib/errors/errors_test.go:23
              |   | testing.tRunner
              |   | 	/usr/local/Cellar/go/1.17.2/libexec/src/testing/testing.go:1259
              |   | runtime.goexit
              |   | 	/usr/local/Cellar/go/1.17.2/libexec/src/runtime/asm_amd64.s:1581
              | Wraps: (2) bar
              | Error types: (1) *withstack.withStack (2) *errutil.leafError
              | (3) baz
              | (1) baz
              | Error types: (1) *errors.errBazType
              | (4) zoo
              | (1) zoo
              | Error types: (1) *errors.errZooType
            Wraps: (2) zoo
            Error types: (1) *errors.multiError (2) *errors.errZooType
        /Users/robertlin/Projects/sourcegraph/sourcegraph/lib/errors/errors_test.go:98: 
            	Error Trace:	errors_test.go:98
            	Error:      	Should be in error chain:
            	            	expected: %!q(**errors.errBazType=0xc00011a0c8)
            	            	in chain: "4 errors occurred:\n\t* foo\n\t* bar\n\t* baz\n\t* zoo"
            	            		"zoo"
            	Test:       	TestMultiError/Append(Append(Append(Append)))
```

File: lib/errors/cockroach.go (hunk 4 above)
Comment (bobheadxi) [reply lib/errors/cockroach.go:56]: I don't think so, since we enforce creating errors with `New` which will attach stacktraces etc - however, from external libraries etc. it will just print whatever those errors indicate

File: internal/observation/observation.go (hunk 1 above)
Comment (bobheadxi) [reply internal/observation/observation.go:219]: Previously with `multierror` that's what it would do I think. I think some code further down checks for a collector and sets to nil if it's empty, so should be safe, if a bit of a odd use of `error` 😄

File: lib/errors/cockroach.go (hunk 4 above)
Comment (camdencheek) [reply lib/errors/cockroach.go:56]: It looks like details is only added with `"%+v`? So if you don't want the full stacktrace, you can just use `%+v`, yes?

**Review by camdencheek — APPROVED** [review]
Re-approving even though I approved before because now `As` should work as expected

File: lib/errors/cockroach.go (hunk 4 above)
Comment (bobheadxi) [reply lib/errors/cockroach.go:56]: you can just use `%v`, or `%s` etc to get the output very similar to `go-multierror` used to get, with two changes:

- single error multi errors print as single errors
- multi error multi errors no long add two extra `\n` at the end

File: lib/errors/group.go (hunk 2 above)
Comment (bobheadxi) [reply lib/errors/group.go:9]: I tried writing a follow-up to this and am not quite sure how to word it, or understand our various usages of parallel utilities enough to adequately describe what unification would look like/mean. If this is something yall are interested could someone open up an issue and link back to this discussion? Thank you! 😁

File: lib/errors/errors.go [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+	errs []error
+}
+
+var _ error = (*multiError)(nil)
+var _ MultiError = (*multiError)(nil)
+
+func combineNonNilErrors(err1 error, err2 error) MultiError {
+	var multi1, multi2 MultiError
+	ok1 := As(err1, &multi1)
+	ok2 := As(err2, &multi2)
```
[the remark this answers is not in the document]
Comment (bobheadxi) [reply lib/errors/errors.go]: https://github.com/sourcegraph/sourcegraph/issues/31497

**Comment by bobheadxi** [comment]
Updated main-dry-run passed: https://buildkite.com/sourcegraph/sourcegraph/builds/132578
